src/java.base/share/classes/jdk/internal/event/SerializationMisdeclarationEvent.java

```diff
@@ -0,0 +1,127 @@
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package jdk.internal.event;
+
+/**
+ * A JFR event for serialization misdeclarations.
+ * This event is mirrored in {@code jdk.jfr.events.SerializationMisdeclarationEvent}
+ * where the metadata for the event is provided with annotations.
+ * Some of the methods are replaced by generated methods when jfr is enabled.
+ * Note that the order of the arguments of the {@link #commit(long,Class,int,String)}
+ * method must be the same as the order of the fields.
+ */
+public class SerializationMisdeclarationEvent extends Event {
+
+    public Class<?> cls;
+    public int kind;
+    public String message;
+
+    /**
+     * Commit a serialization misdeclaration event.
+     * The implementation of this method is generated automatically if jfr is enabled.
+     * The order of the fields must be the same as the parameters in this method.
+     * {@code commit(..., Class, int, String)}
+     *
+     * @param start         timestamp of the start of the operation
+     * @param cls           the affected class
+     * @param kind          the kind of misdeclaration event
+     * @param message       the specific event message
+     */
+    public static void commit(long start, Class<?> cls, int kind, String message) {
+        // Generated by JFR
+    }
+
+    /**
+     * Determine if an event should be emitted.  The duration of the operation
+     * must exceed some threshold in order to commit the event.  The implementation
+     * of this method is generated automatically if jfr is enabled.
+     *
+     * @param duration  time in nanoseconds to complete the operation
+     * @return  true if the event should be commited
+     */
+    public static boolean shouldCommit(long duration) {
+        // Generated by JFR
+        return false;
+    }
+
+    /**
+     * Determine if this kind of event is enabled.  The implementation
+     * of this method is generated automatically if jfr is enabled.
+     *
+     * @return whether serialization misdeclaration events are enabled
+     */
+    public static boolean enabled() {
+        // Generated by JFR
+        return false;
+    }
+
+    /**
+     * Fetch the current timestamp in nanoseconds.  This method is used
+     * to determine the start and end of an operation.  The implementation
+     * of this method is generated automatically if jfr is enabled.
+     *
+     * @return  the current timestamp value
+     */
+    public static long timestamp() {
+        // Generated by JFR
+        return 0L;
+    }
+
+    /*
+     * These constants are not final on purpose.
```

Comment:
Just curious - what's the purpose? I don't see them being changed/updated anywhere (not even in tests).

Reply:
Declaring a `public static int` field `final` means that the value is then stored in the client class. If a value is changed, then the client needs to be recompiled as well, but this is not enforced by the language, so it might lead to inconsistencies.

The clean solution would be to use an enum class, but that's not supported by JFR.

Reply:
Not having them final is a quite smelly. I would suggest to make them final with a comment that new values can be added but that existing values must not be changed.